### ISSUE TITLE
Feature: flexible merging of rows via SQL query

### ISSUE THREAD

**Issue by arctica** [ClickHouse/ClickHouse#4419]
There are currently 6 specialized *MergeTree table engines which have different functionality when it comes to merging of rows. The simplest one, SummingMergeTree for example just sums numeric values while AggregatingMergeTree offers more flexibility in that one can specify different aggregating functions for each column. The Replacing/Collapsing/VersionCollapsing engines allow one to replace rows with newer values.

While powerful, these engines are still quite specific in their usecase and therefor limited in what they can do. One can for example not implement an Exponential Moving Average (EMA) where newer rows are assigned a bigger weight compared to older rows when computing the avg() as far as I could tell.

I propose a feature that would enable complete flexibility when it comes to how multiple rows should be merged by letting the user specify a SQL query which can use any aggregate functions together with arbitrary data transformations.

This could be implemented via a SQL query specified as a table parameter

```
CREATE TABLE test (ticker String, ts DateTime, value UInt64) ENGINE = MergeTree
ORDER BY (ticker, ts)
AGGREGATE BY SELECT ticker, max(ts), avg(value * (3600 - max(now() - ts, 3600)) / 3600) GROUP BY ticker HAVING value > 0
```

The above example would calculate some weighted moving average for the value column relative to how old it is and only retains rows for which the value is bigger than zero.

This could also be used to implement data retention policies.

There are two negatives that I can see with this approach:
1. One can never assume that the part is "final" even if no new rows had been written since the last merge. This is demonstrated also by the above example. This can never be resolved and one would have to run OPTIMIZE FINAL queries if one knows that according to the specified logic, a part could be further aggregated. In practice, I don't think this would be a real issue.
2. Performance of this approach is most likely much worse than with the specialized table engines that currently exist. I assume that the query has to process all rows of the parts that are being merged with intermittent results in memory before it can write the results to disk unless there can be some clever optimizations applied. The performance penalty is clearly outweighed though by the tremendous amount of flexibility that this feature offers as many aggregation schemes simply can't be realized with the specialized engines. It would not replace the specialized engines.

BTW it would be good if one could specify a max part size to avoid the aggregation query consuming huge amounts of memory for really big parts. Maybe such a limit already exists but I could not find such a setting.

**Comment by arctica** [comment]
Actually I think the proposed feature would also obsolete #4413 because one could have a table with just a few columns in the primary/sorting key and use the AGGREGATE BY query to sort the rows however one desires. And because the columns are not part of the primary/sort key, one is free to add or remove columns as one sees fit. Updating the AGGREGATE BY query should be seen as a metadata-only change so would be "free" as well. Whenever one adds/removes a column, one can also adjust the AGGREGATE BY query accordingly.

> CREATE TABLE test (a String, v UInt32) ENGINE = MergeTree() PRIMARY KEY a
> AGGREGATE BY SELECT a, sum(v) GROUP BY a ORDER BY a

> ALTER TABLE test ADD COLUMN b String AFTER a,
> MODIFY AGGREGATE BY SELECT a, b, sum(v) GROUP BY a, b ORDER BY a, b

The AGGREGATE BY query would not guarantee that data is sorted already at any moment in time but that totally in line with the low guarantees due to sharding and the nature of MergeTree with different data parts. This is in contrast to PRIMARY KEY and ORDER BY key which guarantee order within a part but come with heavy restrictions.

Alternatively instead of AGGREGATE BY one might also consider MERGE BY. Maybe the SELECT prefix could be omitted as well.

**Comment by amosbird** [comment]
It's likely to be superceded by aggregate projections in https://github.com/ClickHouse/ClickHouse/pull/20202

**Comment by arctica** [comment]
@amosbird projections seem like a very useful and interesting feature because they solve the (big) limitations of materialized views but I'm not sure if they could be used to serve the same purpose as my suggestion.

The feature I describe can reduce storage costs a lot because it aggregates the data of the table during merges of parts. If I understood the proposed projects feature correctly, those projections will actually increase storage costs because they store the projected data in addition to the original data. Also it seems like if the original part is deleted, the data will be removed from the projection as well? The story around data mutations seems not clear for projections at this point.

Also projections don't seem to be able to be chained. What I mean is that raw data is inserted in table T, then projection P1 takes the raw data and aggregates it and projection P2 takes the data from P1 and aggregates it further.

Real world example: you have a web analytics services where you want to keep raw click events for the past 24h, then aggregated numbers per hour for the past month, then aggregated numbers per day for the past year.

The original motivation for my proposal was similar to the above example: allow for a table which stores a log of events and stores aggregate data over time. The older the data, the less granular it would be. This is actually quite tricky to achieve in ClickHouse because:

1. Materialized Views have too many severe limitations like only aggregating data on an inserted block level.
2. ClickHouse has no transactions nor way to write data to more than one table atomically.
3. The built-in aggregated table engines are very limited in what they can do.

The proposed projections come close but not quite imho.

**Comment by amosbird** [comment]
> those projections will actually increase storage costs because they store the projected data in addition to the original data. Also it seems like if the original part is deleted, the data will be removed from the projection as well?

It's currently possible by using CollapsingMergeTree with default `sign = 1`. We wil add ProjectionMergeTree if necessary.


> Also projections don't seem to be able to be chained. What I mean is that raw data is inserted in table T, then projection P1 takes the raw data and aggregates it and projection P2 takes the data from P1 and aggregates it further.

You can build projection P2 directly from base table.

>  The older the data, the less granular it would be.

This is the motivation of ProjectionMergeTree with specialized TTL rules. It's still under consideration.

**Comment by arctica** [comment]
> It's currently possible by using CollapsingMergeTree with default `sign = 1`. We wil add ProjectionMergeTree if necessary.

Yes, as I outlined in the beginning of my proposal, there are many specialized table engines which all solve one usecase but none can do it all and you can't really combine them. Table engine A can solve problem X and B can solve Y but none can solve X and Y.

> You can build projection P2 directly from base table.

It would be potentially orders of magnitude faster to enable a projection from a projection. Imagine there are 1B rows in a table which the projection P1 aggregates into 1M rows. P2 aggregates the data into 1K rows. If P2 could read from P1 then they'd need to process in total 1.001B rows instead of 2B.

> > The older the data, the less granular it would be.
> 
> This is the motivation of ProjectionMergeTree with specialized TTL rules. It's still under consideration.

Making projections their own table engine would be much better in my opinion as you'd get the flexibility I outlined. You could have a MergeTree table raw_events and then have a projection1 table and from there add a projection2 table.

The trickiest part I guess is to define when a reculcuation on a projection should be performed and how to handle deletes if one wants to keep the data in a projection even after it is deleted from the source table.

1. When the source table has a mutation (new part/block, merge, delete)
2. When two parts of the projection table are merged

You also get a much simpler implementation because you don't have to analyze SELECT queries to figure out what should be read from where (main table, projection1, projection2?). I imagine this can be extremely tricky with many edge cases that are super hard or even impossible to figure out. With explicit tables as projections, the user has to do a bit more work in the query has he potentially has to JOIN but that's quite easy and has very clear semantics.

This would be a hybrid between my proposal which only aggregates its own data when merging parts and your projections proposal. Materialized Views done right. Let me know what you think.

I saw on the roadmap that atomic inserts on multiple tables is being worked on which would enable data consistency for ProjectionMergeTree and source tables.

**Comment by bash-spbu** [comment]
I guess I have a use case for this one described here: https://github.com/ClickHouse/ClickHouse/discussions/66965

In addition to the features proposed here, this one can also be solved with either of the following features:
- UDAF (user defined aggregate functions)
- New column specifier in the addition to the `MATERIALIZED | DEFAULT | ALIAS` -- `COMPUTABLE` which is like `MATERIALIZED` but with the only exception that it is computed not only on the insert to the table, but also on everty parts merge. This seems to be the easiest to support and would allow to compute one column aggregation based on some expression which could use results of other column aggregations.
